scene/resources/mesh.cpp

```diff
@@ -954,127 +933,93 @@ void _fix_array_compatibility(const Vector<uint8_t> &p_src, uint32_t p_old_forma
+							dst[1] = (int16_t)CLAMP(src[1] / 127.0f * 32767, -32768, 32767);
 						}
-						src_offset += sizeof(int8_t) * 2;
+						src_offset += sizeof(int16_t) * 2;
 					} else {
 						for (uint32_t i = 0; i < p_elements; i++) {
 							const int16_t *src = (const int16_t *)&src_vertex_ptr[i * src_vertex_stride + src_offset];
-							uint32_t *dst = (uint32_t *)&dst_vertex_ptr[i * dst_vertex_stride + dst_offsets[Mesh::ARRAY_NORMAL]];
-							const Vector2 src_vec(src[0] / 32767.0f, src[1] / 32767.0f);
-
-							const Vector3 res = _oct_to_norm(src_vec) * Vector3(0.5, 0.5, 0.5) + Vector3(0.5, 0.5, 0.5);
-							*dst = 0;
-							*dst |= CLAMP(int(res.x * 1023.0f), 0, 1023);
-							*dst |= CLAMP(int(res.y * 1023.0f), 0, 1023) << 10;
-							*dst |= CLAMP(int(res.z * 1023.0f), 0, 1023) << 20;
+							int16_t *dst = (int16_t *)&dst_vertex_ptr[i * dst_vertex_stride + dst_offsets[Mesh::ARRAY_NORMAL]];
+
+							dst[0] = src[0];
+							dst[1] = src[1];
 						}
 						src_offset += sizeof(int16_t) * 2;
 					}
 				} else { // No Octahedral compression
 					if (p_old_format & OLD_ARRAY_COMPRESS_NORMAL) {
-						const float multiplier = 1.f / 127.f * 1023.0f;
-
 						for (uint32_t i = 0; i < p_elements; i++) {
 							const int8_t *src = (const int8_t *)&src_vertex_ptr[i * src_vertex_stride + src_offset];
-							uint32_t *dst = (uint32_t *)&dst_vertex_ptr[i * dst_vertex_stride + dst_offsets[Mesh::ARRAY_NORMAL]];
+							const Vector3 original_normal(src[0], src[1], src[2]);
+							Vector2 res = original_normal.octahedron_encode();
 
-							*dst = 0;
-							*dst |= CLAMP(int(src[0] * multiplier), 0, 1023);
-							*dst |= CLAMP(int(src[1] * multiplier), 0, 1023) << 10;
-							*dst |= CLAMP(int(src[2] * multiplier), 0, 1023) << 20;
+							uint16_t *dst = (uint16_t *)&dst_vertex_ptr[i * dst_vertex_stride + dst_offsets[Mesh::ARRAY_NORMAL]];
+							dst[0] = (uint16_t)CLAMP(res.x * 65535, 0, 65535);
+							dst[1] = (uint16_t)CLAMP(res.y * 65535, 0, 65535);
 						}
-						src_offset += sizeof(uint32_t);
+						src_offset += sizeof(uint16_t) * 2;
 					} else {
 						for (uint32_t i = 0; i < p_elements; i++) {
 							const float *src = (const float *)&src_vertex_ptr[i * src_vertex_stride + src_offset];
-							uint32_t *dst = (uint32_t *)&dst_vertex_ptr[i * dst_vertex_stride + dst_offsets[Mesh::ARRAY_NORMAL]];
+							const Vector3 original_normal(src[0], src[1], src[2]);
+							Vector2 res = original_normal.octahedron_encode();
 
-							*dst = 0;
-							*dst |= CLAMP(int(src[0] * 1023.0), 0, 1023);
-							*dst |= CLAMP(int(src[1] * 1023.0), 0, 1023) << 10;
-							*dst |= CLAMP(int(src[2] * 1023.0), 0, 1023) << 20;
+							uint16_t *dst = (uint16_t *)&dst_vertex_ptr[i * dst_vertex_stride + dst_offsets[Mesh::ARRAY_NORMAL]];
+							dst[0] = (uint16_t)CLAMP(res.x * 65535, 0, 65535);
+							dst[1] = (uint16_t)CLAMP(res.y * 65535, 0, 65535);
 						}
-						src_offset += sizeof(float) * 3;
+						src_offset += sizeof(uint16_t) * 2;
```

Comment:
you changed it here, too. `sizeof(float) = 4`, src is `float`, we're getting `src[0]`, `src[1]`, and `src[2]`. 12 bytes total. You're only incrementing by 4 bytes.